File: week-5/is-graph-bipartite/is-graph-bipartite.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        colors = [0 for i in graph]
        for i in range(len(graph)):
            if colors[i] == 0 and not self.dfs(i, graph, colors, 1):
                return False
            
        return True
    
    def dfs(self, i, graph, colors, cur_color):
        if colors[i] != 0:
            return cur_color == colors[i]
        
        colors[i] = cur_color
        for j in graph[i]:
            if not self.dfs(j, graph, colors, -cur_color):
                return False
            
        return True
        

File: week-5/is-graph-bipartite/is-graph-bipartite.js
/**
 * @param {number[][]} graph
 * @return {boolean}
 */
var isBipartite = function(graph) {
    const colors = new Array(graph.length).fill(0);
    for (let i=0; i<graph.length; i++) {
        if (colors[i] === 0 && !dfs(i, graph, colors, 1)) return false
    }
    
    return true;
};

const dfs = (i, graph, colors, cur_color) => {
    if (colors[i] !== 0) return colors[i] === cur_color;
    
    colors[i] = cur_color;
    for (let j of graph[i]) {
        if (!dfs(j, graph, colors, -cur_color)) return false;
    }
    
    return true;
};
